google/cloud/storage/tests/object_read_range_integration_test.cc

```diff
@@ -0,0 +1,165 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/client.h"
+#include "google/cloud/storage/testing/storage_integration_test.h"
+#include "google/cloud/internal/getenv.h"
+#include "google/cloud/testing_util/scoped_environment.h"
+#include "google/cloud/testing_util/status_matchers.h"
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace {
+
+using ::google::cloud::internal::GetEnv;
+using ::google::cloud::testing_util::IsOk;
+
+class ObjectReadRangeIntegrationTest
+    : public ::google::cloud::storage::testing::StorageIntegrationTest {
+ public:
+  ObjectReadRangeIntegrationTest() = default;
+
+ protected:
+  void SetUp() override {
+    bucket_name_ =
+        GetEnv("GOOGLE_CLOUD_CPP_STORAGE_TEST_BUCKET_NAME").value_or("");
+
+    ASSERT_FALSE(bucket_name_.empty());
+  }
+
+  std::string const& bucket_name() const { return bucket_name_; }
+
+ private:
+  std::string bucket_name_;
+};
+
+TEST_F(ObjectReadRangeIntegrationTest, ReadRangeSmall) {
+  StatusOr<Client> client = MakeIntegrationTestClient();
+  ASSERT_STATUS_OK(client);
+
+  auto const contents = LoremIpsum();
+  auto const object_name = MakeRandomObjectName();
+
+  auto insert = client->InsertObject(bucket_name(), object_name, contents,
+                                     IfGenerationMatch(0));
+  ASSERT_THAT(insert, IsOk());
+  EXPECT_THAT(contents.size(), insert->size());
+
+  auto size = static_cast<std::int64_t>(insert->size());
+
+  // Read several small portions of the object, expecting specific results.
+  struct Test {
+    std::int64_t begin;
+    std::int64_t end;
+    std::string expected;
+  } cases[] = {
+      {0, 1, contents.substr(0, 1)},
+      {4, 8, contents.substr(4, 4)},
+      {0, size, contents},
+  };
+
+  for (auto const& test : cases) {
+    SCOPED_TRACE("Testing range [" + std::to_string(test.begin) + "," +
+                 std::to_string(test.end) + ")");
+    auto reader = client->ReadObject(bucket_name(), object_name,
+                                     ReadRange(test.begin, test.end));
+    auto actual = std::string{std::istreambuf_iterator<char>(reader), {}};
+    EXPECT_THAT(reader.status(), IsOk());
```

Comment:
Should you do this before constructing `actual`?

Reply:
I am like 99.4% certain that it is better to check `.status()` after constructing `actual`:

* For small objects like these it does not matter, a failed download fails immediately, and `actual` is empty because the iterator stops immediately,
* For larger objects the download may start successfully (at which point `.status().code()` is `kOkay`) and fail by the end, and once it fails the `.status()` never changes.

Reply:
Okie dokie.  (2) sounds good, and if the `string` construction in (1) works on a failed reader then that sounds good too.  Thanks.